src/layers/domain/repository/device_repository/v2.py

```diff
@@ -0,0 +1,190 @@
+from domain.repository.keys import TableKeys
+from domain.repository.marshall import marshall, marshall_value, unmarshall
+from domain.repository.repository.v2 import Repository
+from domain.repository.transaction import (
+    ConditionExpression,
+    TransactionStatement,
+    TransactItem,
+)
+
+
+def asdict(obj) -> dict:
+    return _asdict(obj, recurse=False)
+
+
+def _dynamodb_update_expression(updated_fields: dict):
+    expression_attribute_names = {}
+    expression_attribute_values = {}
+    update_clauses = []
+
+    for field_name, value in updated_fields.items():
+        field_name_placeholder = f"#{field_name}"
+        field_value_placeholder = f":{field_name}"
+
+        update_clauses.append(f"{field_name_placeholder} = {field_value_placeholder}")
+        expression_attribute_names[field_name_placeholder] = field_name
+        expression_attribute_values[field_value_placeholder] = marshall_value(value)
+
+    update_expression = "SET " + ", ".join(update_clauses)
+
+    return dict(
+        UpdateExpression=update_expression,
+        ExpressionAttributeNames=expression_attribute_names,
+        ExpressionAttributeValues=expression_attribute_values,
+    )
+
+
+def _device_primary_keys(device_id: str, device_keys: list[DeviceKey]) -> list[dict]:
+    """
+    Generates all the fully marshalled (i.e. {"pk": {"S": "123"}} DynamoDB
+    primary keys (i.e. pk + sk) for the provided Device. This is one primary key
+    plus an additional primary key for every value of Device.keys
+    """
+    root_primary_key = marshall(
+        pk=TableKeys.DEVICE.key(device_id), sk=TableKeys.DEVICE.key(device_id)
+    )
+    non_root_primary_keys = [
+        marshall(pk=pk, sk=pk)
+        for pk in (TableKeys.DEVICE.key(k.key_type, k.key_value) for k in device_keys)
+    ]
+    return [root_primary_key] + non_root_primary_keys
+
+
+def handle_update_device(
+    table_name: str, id: str, keys: list[DeviceKey], data: dict
+) -> list[TransactItem]:
+    primary_keys = _device_primary_keys(device_id=id, device_keys=keys)
+    update_expression = _dynamodb_update_expression(updated_fields=data)
+    update_statement = partial(
+        TransactionStatement,
+        TableName=table_name,
+        ConditionExpression=ConditionExpression.MUST_EXIST,
+        **update_expression,
+    )
+    transact_items = [
+        TransactItem(Update=update_statement(Key=key)) for key in primary_keys
+    ]
+    return transact_items
+
+
+def handle_create_device(
+    table_name: str, key_parts: tuple[str], device_data: dict
+) -> TransactItem:
+    key = TableKeys.DEVICE.key(*key_parts)
+    return TransactItem(
+        Put=TransactionStatement(
+            TableName=table_name,
+            Item=marshall(pk=key, sk=key, root=True, **device_data),
+            ConditionExpression=ConditionExpression.MUST_NOT_EXIST,
+        )
+    )
+
+
+def handle_delete_device(table_name: str, key_parts: tuple[str]) -> TransactItem:
```

Comment:
I dont think we had a delete device mechanism before so I suppose this comment doesnt matter too much, but i dont think this will this go and delete all the extra rows that belong to that "device" will it?

Reply:
it deletes a given single device row by key - so the intention is to delete either a tag-indexed or a key-indexed device - not the root device. If we ever accidentally used this function actually hard delete root devices that would be caught by integration tests, as there all tests currently indirectly check that no device is hard deleted anyway

Reply:
I think you wernt for device_index on the other PR which works!